sycl/doc/extensions/DataFlowPipes/data_flow_pipes.asciidoc

```diff
@@ -0,0 +1,650 @@
+
+Uses of pipes declared within a function call that is invoked in a concurrent way are not common, but are legal.
+
+Users can privatize the pipe type by templating the function, if they do require independent pipes for concurrent calls to the function.  For example:
+
+[source,c++,TemplatedFunctionCallExample,linenums]
+----
+template <int id>
+void pipe_memcpy(int* dest, const int* src, size_t n) {
+  constexpr int N = 10;  // n <= N.  No checking here for simplicity
+  using mypipe = pipe<class local, int, N>;
+  for (size_t i = 0; i < n; ++i) mypipe::write(src[i]);
+  for (size_t i = 0; i < n; ++i) dest[i] = mypipe::read();
+}
+
+...
+
+pipe_memcpy<1>(d1, s1, N);
+pipe_memcpy<2>(d2, s2, N);
+----
+
+The above example function `pipe_memcpy()` could alternatively be templated on the pipe type (first template argument of the pipe class specialization), amongst other possibilities.
+
+Automated mechanisms are possible to provide uniquification across calls, and could be exposed through a wrapper or library.
+
+== [[warnings]]Required warning messages needing compiler support
+
+. Warning if two pipes are found within the translation unit that have an identical first template argument, and differ only in one or more of the following template arguments.
+
+== Issues
+
+. Should a warning be required if there is no forward declaration of a pipe type (e.g. declared within an expression)?  More specifically, should we add a required compiler warning/error if a pipe name type (first template argument of pipe type) is declared at a scope local to the pipe type declaration?  This would result in feedback to the user if they didn't declare the type at some scope larger than the pipe declaration, which is possibly a bug that will result in misconnected or unconnected pipes.
++
+--
+*RESOLUTION*: Not resolved.  Looking for input, because this is a valid design pattern in some cases.
+--
+
+. Arbitration is allowed by default (more than one read or write endpoint) within a single kernel.  Should there be an additional pipe template parameter to disable arbitration, as part of the type?  Downsides are that restriction as part of the type requires compiler support, since the pipe and read/write methods are stateless, and adding additional parameters to the type increases likelihood of accidentally creating two pipes with slightly different parameterizations.
++
+--
+*RESOLUTION*: Resolved.  Not adding template parameter in this version because want to minimize parameters of the type.  But open to further input.
+--
+
+. Pipe types effectively link globally.  There is no way to scope a type to only apply in a restricted region of code, with the same type elsewhere forming a different pipe.  Namespaces can make this very manageable, and in fact the global linking is a key feature, but adding as an issue to make sure that this is well understood.
```

Comment:
Is it not why unnamed namespace in C++11 have been introduced for?
https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces

Reply:
Yes, agreed, but that still applies at relatively global scope (namespace scope).  I changed the wording to try to capture better.